src/librustc/hir/mod.rs

```diff
@@ -444,71 +446,54 @@ pub enum TraitBoundModifier {
 pub type TyParamBounds = HirVec<TyParamBound>;
 
 #[derive(Clone, PartialEq, Eq, RustcEncodable, RustcDecodable, Hash, Debug)]
-pub struct TyParam {
-    pub name: Name,
+pub enum GenericParamKind {
+    /// A lifetime definition, eg `'a: 'b + 'c + 'd`.
+    Lifetime {
+        /// Either "'a", referring to a named lifetime definition,
+        /// or "" (aka keywords::Invalid), for elision placeholders.
+        ///
+        /// HIR lowering inserts these placeholders in type paths that
+        /// refer to type definitions needing lifetime parameters,
+        /// `&T` and `&mut T`, and trait objects without `... + 'a`.
+        name: LifetimeName,
+        bounds: HirVec<Lifetime>,
+        // Indicates that the lifetime definition was synthetically added
+        // as a result of an in-band lifetime usage like:
+        // `fn foo(x: &'a u8) -> &'a u8 { x }`
+        in_band: bool,
+        // We keep a `Lifetime` around for now just so we can `visit_lifetime`.
+        lifetime: Lifetime,
+    },
+    Type {
+        name: Name,
+        bounds: TyParamBounds,
+        default: Option<P<Ty>>,
+        synthetic: Option<SyntheticTyParamKind>,
+        attrs: HirVec<Attribute>,
+    }
+}
+
+#[derive(Clone, PartialEq, Eq, RustcEncodable, RustcDecodable, Hash, Debug)]
+pub struct GenericParam {
     pub id: NodeId,
-    pub bounds: TyParamBounds,
-    pub default: Option<P<Ty>>,
     pub span: Span,
     pub pure_wrt_drop: bool,
```

Comment:
Huh, it's desugared this early?

Reply:
Desugared from what?

Reply:
It comes from a `#[may_dangle]` attribute.